src/analysis/schwartz.lean

```diff
@@ -0,0 +1,400 @@
+f.decay' k n
+
+/-- Every Schwartz function is smooth. -/
+lemma smooth (f : 𝓢(E, F)) : cont_diff ℝ ⊤ f := f.smooth'
+
+@[ext] lemma ext {f g : 𝓢(E, F)} (h : ∀ x, (f : E → F) x = g x) : f = g := fun_like.ext f g h
+
+section aux
+
+lemma bounds_nonempty (k n : ℕ) (f : 𝓢(E, F)) :
+  ∃ (c : ℝ), c ∈ {c : ℝ | 0 ≤ c ∧ ∀ (x : E), ∥x∥^k * ∥iterated_fderiv ℝ n f x∥ ≤ c} :=
+let ⟨M, hMp, hMb⟩ := f.decay k n in ⟨M, le_of_lt hMp, hMb⟩
+
+lemma bounds_bdd_below (k n : ℕ) (f : 𝓢(E, F)) :
+  bdd_below { c | 0 ≤ c ∧ ∀ x, ∥x∥^k * ∥iterated_fderiv ℝ n f x∥ ≤ c } :=
+⟨0, λ _ ⟨hn, _⟩, hn⟩
+
+lemma decay_add_le_aux (k n : ℕ) (f g : 𝓢(E, F)) (x : E) :
+  ∥x∥^k * ∥iterated_fderiv ℝ n (f+g) x∥ ≤
+  ∥x∥^k * ∥iterated_fderiv ℝ n f x∥
+  + ∥x∥^k * ∥iterated_fderiv ℝ n g x∥ :=
+begin
+  rw ←mul_add,
+  refine mul_le_mul_of_nonneg_left _ (by positivity),
+  convert norm_add_le _ _,
+  refine iterated_fderiv_add_apply _ _,
+  { exact f.smooth.of_le (le_of_lt $ with_top.coe_lt_top _) },
+  { exact g.smooth.of_le (le_of_lt $ with_top.coe_lt_top _) },
+end
+
+variables [is_R_or_C 𝕜] [normed_space 𝕜 F]
+variables [semiring R] [module R 𝕜] [module R F] [smul_comm_class ℝ R F]
+variables [has_continuous_const_smul R F] [is_scalar_tower R 𝕜 F]
+
+lemma decay_smul_aux (k n : ℕ) (f : 𝓢(E, F)) (c : R) (x : E) :
+  ∥x∥ ^ k * ∥iterated_fderiv ℝ n (c • f) x∥ =
+  ∥c • (1 : 𝕜)∥ * ∥x∥ ^ k * ∥iterated_fderiv ℝ n f x∥ :=
+begin
+  nth_rewrite 2 mul_comm,
+  rw mul_assoc,
+  congr,
+  rw iterated_fderiv_const_smul_apply ,
+  { rw ←smul_one_smul 𝕜 c,
+    rw norm_smul,
+    apply_instance },
+  { exact f.smooth.of_le (le_of_lt $ with_top.coe_lt_top _) },
+end
+
+lemma decay_neg_aux (k n : ℕ) (f : 𝓢(E, F)) (x : E) :
+  ∥x∥ ^ k * ∥iterated_fderiv ℝ n (-f) x∥ = ∥x∥ ^ k * ∥iterated_fderiv ℝ n f x∥ :=
+begin
+  nth_rewrite 3 ←norm_neg,
+  congr,
+  exact iterated_fderiv_neg_apply,
+end
+
+variables [normed_space ℂ F]
+
+lemma decay_smul_aux' (k n : ℕ) (f : 𝓢(E, F)) (c : ℂ) (x : E) :
+  ∥x∥ ^ k * ∥iterated_fderiv ℝ n (c • f) x∥ =
+  ∥c∥ * ∥x∥ ^ k * ∥iterated_fderiv ℝ n f x∥ :=
+begin
+  nth_rewrite 2 mul_comm,
+  rw mul_assoc,
+  congr,
+  rw iterated_fderiv_const_smul_apply ,
+  { rw norm_smul },
+  { exact f.smooth.of_le (le_of_lt $ with_top.coe_lt_top _) },
+end
+
+end aux
+
+/-! ### Algebraic properties -/
+
+section smul
+
+variables [normed_space ℂ F]
+variables [semiring R] [module R ℂ] [module R F] [smul_comm_class ℝ R F]
+variables [has_continuous_const_smul R F] [is_scalar_tower R ℂ F]
+variables [semiring R'] [module R' ℂ] [module R' F] [smul_comm_class ℝ R' F]
+variables [has_continuous_const_smul R' F] [is_scalar_tower R' ℂ F]
+
+-- Note that we define the scalar multiplication only in the case that `F` is a vector space
+-- over `ℂ`. The reason for this is that the type-system cannot infer instances if we were to
+-- replace `ℂ` by `[is_R_or_C 𝕜]`. This is mathemically no problem, because the usual Schwartz
+-- space is `𝓢(E, ℂ)` and the space `𝓢(E, ℝ)` is never used in mathematics.
```

Comment:
I am not happy with this. Can't you just assume that `F` is a vector space over `R` with the right tower properties between `R` and `ℝ` , and not mention `ℂ` at all, to have something that would work simultaneously for `R = ℝ` or `R = ℂ`?

Reply:
I tried and did not get it to work using `is_scalar_tower`. This part has been a major pain for me and I basically ran out of ideas how to get it to work. If you have any ideas how to get rid of the complex numbers here, I would be happy.